planner/core/find_best_task.go

```diff
@@ -1520,13 +1520,28 @@ func (ds *DataSource) convertToIndexScan(prop *property.PhysicalProperty,
 	}
 	if candidate.isMatchProp {
 		cop.keepOrder = true
-		// IndexScan on partition table can't keep order.
 		if ds.tableInfo.GetPartitionInfo() != nil {
-			return invalidTask, nil
+			// Add sort items for index scan for merge-sort operation between partitions.
+			byItems := make([]*util.ByItems, 0, len(prop.SortItems))
+			for _, si := range prop.SortItems {
+				byItems = append(byItems, &util.ByItems{
+					Expr: si.Col,
+					Desc: si.Desc,
+				})
+			}
+			cop.indexPlan.(*PhysicalIndexScan).ByItems = byItems
 		}
 		if cop.tablePlan != nil && !ds.tableInfo.IsCommonHandle {
 			col, isNew := cop.tablePlan.(*PhysicalTableScan).appendExtraHandleCol(ds)
 			cop.extraHandleCol = col
+			if ds.tableInfo.GetPartitionInfo() != nil && !is.Index.Global {
+				is.Columns = append(is.Columns, model.NewExtraPhysTblIDColInfo())
+				is.schema.Append(&expression.Column{
+					RetType:  types.NewFieldType(mysql.TypeLonglong),
+					UniqueID: ds.ctx.GetSessionVars().AllocPlanColumnID(),
+					ID:       model.ExtraPhysTblID,
+				})
+			}
```

Comment:
Why do we need this? Does this conflict with `(is *PhysicalIndexScan) initSchema`?

Reply:
No, that case is to add a partition id for the global index(It's decoded from the value of the global index). Here we add an extra partition id for non-global index which is decoded from the key instead of the value.

Reply:
`ExtraPhysTblID` will let TiKV return PID in chunk. IndexWorker will use the pid to create a partitionHandle for tableWorker found the handle belongs to which PID.